File: app/src/main/java/com/task/temponewstask/utils/AppPrefs.java
package com.task.temponewstask.utils;

import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.google.gson.Gson;
import com.task.temponewstask.AppClass;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

public class AppPrefs {
    private static AppPrefs appPrefs;
    private static final String REFRESH_TOKEN = "REFRESH_TOKEN";
    private SharedPreferences pref;
    private SharedPreferences.Editor editor;

    public AppPrefs() {
        pref = PreferenceManager.getDefaultSharedPreferences(AppClass.Companion.getInstance());
        editor = pref.edit();
    }

    public void saveRefreshToken(String token) {
        editor.putString(REFRESH_TOKEN, token);
        editor.apply();
    }

    public String getRefreshToken() {
        return pref.getString(REFRESH_TOKEN, null);
    }

    public static AppPrefs get() {
        if (appPrefs == null)
            appPrefs = new AppPrefs();
        return appPrefs;
    }

    public <T> void saveAsJson(T data, Class<T> clazz) {
        editor.putString(clazz.getSimpleName(), new Gson().toJson(data));
        editor.apply();
    }

    public <T> T fromJson(Class<T> clazz) {
        String json = pref.getString(clazz.getSimpleName(), "");
        return new Gson().fromJson(json, clazz);
    }
}

File: app/src/main/java/com/task/temponewstask/utils/AppPrefs.kt
package com.task.temponewstask.utils

import android.content.SharedPreferences
import android.preference.PreferenceManager
import com.google.gson.Gson
import com.task.temponewstask.AppClass.Companion.instance

class AppPrefs {
    private val pref: SharedPreferences
    private val editor: SharedPreferences.Editor
    fun saveRefreshToken(token: String?) {
        editor.putString(REFRESH_TOKEN, token)
        editor.apply()
    }

    val refreshToken: String?
        get() = pref.getString(REFRESH_TOKEN, null)

    fun <T> saveAsJson(data: T, clazz: Class<T>) {
        editor.putString(clazz.simpleName, Gson().toJson(data))
        editor.apply()
    }

    fun <T> fromJson(clazz: Class<T>): T {
        val json = pref.getString(clazz.simpleName, "")
        return Gson().fromJson(json, clazz)
    }

    companion object {
        private var appPrefs: AppPrefs? = null
        private const val REFRESH_TOKEN = "REFRESH_TOKEN"
        fun get(): AppPrefs? {
            if (appPrefs == null) appPrefs = AppPrefs()
            return appPrefs
        }
    }

    init {
        pref = PreferenceManager.getDefaultSharedPreferences(instance)
        editor = pref.edit()
    }
}